pkgs/build-support/setup-hooks/make-wrapper.sh

```diff
@@ -30,88 +30,96 @@ assertExecutable() {
+            local value="$4"
+            shift 4
             if test -n "$value"; then
                 if test "$p" = "--suffix"; then
                     echo "export $varName=\$$varName\${$varName:+${separator@Q}}${value@Q}" >> "$wrapper"
                 else
                     echo "export $varName=${value@Q}\${$varName:+${separator@Q}}\$$varName" >> "$wrapper"
                 fi
             fi
-        elif [[ "$p" == "--suffix-each" ]]; then
-            varName="${params[$((n + 1))]}"
-            separator="${params[$((n + 2))]}"
-            values="${params[$((n + 3))]}"
-            n=$((n + 3))
+            ;;
+        --suffix-each)
+            local varName="$2"
+            local separator="$3"
+            local values="$4"
+            shift 4
             for value in $values; do
                 echo "export $varName=\$$varName\${$varName:+$separator}${value@Q}" >> "$wrapper"
             done
-        elif [[ ("$p" == "--suffix-contents") || ("$p" == "--prefix-contents") ]]; then
-            varName="${params[$((n + 1))]}"
-            separator="${params[$((n + 2))]}"
-            fileNames="${params[$((n + 3))]}"
-            n=$((n + 3))
+            ;;
+        --suffix-contents | --prefix-contents)
+            local varName="$2"
+            local separator="$3"
+            local fileNames="$4"
+            shift 4
             for fileName in $fileNames; do
-                contents="$(cat "$fileName")"
+                contents="$(< "$fileName")"
                 if test "$p" = "--suffix-contents"; then
                     echo "export $varName=\$$varName\${$varName:+$separator}${contents@Q}" >> "$wrapper"
                 else
                     echo "export $varName=${contents@Q}\${$varName:+$separator}\$$varName" >> "$wrapper"
                 fi
             done
-        elif [[ "$p" == "--add-flags" ]]; then
-            flags="${params[$((n + 1))]}"
-            n=$((n + 1))
-            flagsBefore="$flagsBefore $flags"
-        elif [[ "$p" == "--argv0" ]]; then
-            argv0="${params[$((n + 1))]}"
-            n=$((n + 1))
-        else
-            die "makeWrapper doesn't understand the arg $p"
-        fi
+            ;;
+        --add-flags)
+            local flags="$2"
+            shift 2
+            flagsBefore+=($flags)
+            ;;
+        --argv0)
+            local argv0="$2"
+            shift 2
+            ;;
+        *)
+            die "makeWrapper doesn't understand the arg ${1@Q}"
+    esac
     done
 
     # Note: extraFlagsArray is an array containing additional flags
     # that may be set by --run actions.
     # Silence warning about unexpanded extraFlagsArray:
     # shellcheck disable=SC2016
-    echo exec ${argv0:+-a \"$argv0\"} \""$original"\" \
-         "$flagsBefore" '"${extraFlagsArray[@]}"' '"$@"' >> "$wrapper"
+    printf "%s" "exec ${argv0:+-a ${argv0@Q}} ${original@Q}" >> "$wrapper"
+    printWords "${flagsBefore[@]}" >> "$wrapper"
```

Comment:
This quotes, unlike before.

Reply:
`printWords` is `echo -n` with a trailing space, there is no need for it here, just put `"${flagsBefore[@]}"` instead of `"$flagsBefore"`.

Reply:
`echo` is better than `printf` at inserting spaces between arguments, please remove `printf` and print the whole line with one `echo` as before.

Reply:
Ah I thought printWords used `%q`; ironic as I wrote it. Will go back to echo; should we use `@Q` here too?

Reply:
No, the values of add-flags must be inserted as text to be evaluated when the wrapper runs.